pio/irq/main.c

```diff
@@ -0,0 +1,174 @@
+    // ISR would execute only once when simultaneous interrupts happen instead
+    // of being executed for each interrupt individually.
+    // For other applications, make sure to check the correct pio (pio0/pio1)
+    irq_flags = pio0_hw->irq;
+    // Clear the flags since we've saved them to check later.
+    hw_clear_bits(&pio0_hw->irq, irq_flags);
+
+    // alternatively you could use pio_interrupt_get(pio0, #) to check a specific
+    // irq flag.
+}
+
+/**
+ * Lets the pico know that we want it to notify us of the PIO ISRs.
+ * @note in simple_irq.pio we enable irq0. This tells the state machine
+ *       to send the ISRs to the core, we still need to tell the core
+ *       to send them to our program.
+ */
+void enable_pio_irqs() {
+    // Set the function that will be called when the PIO IRQ comes in.
+    irq_set_exclusive_handler(PIO0_IRQ_0, simple_irq_handler);
+
+    // Once that function is set, we can go ahead and allow the interrupts
+    // to come in. You want to set the function before enabling the interrupt
+    // just in case. The docs say if an IRQ comes in and there's no handler
+    // then it will work like a breakpoint, which seems bad.
+    irq_set_enabled(PIO0_IRQ_0, true);
+}
+
+/**
+ * Loads simple_irq pio program into PIO memory
+ */
+void load_pio_programs() {
+    PIO pio = pio0;
+
+    // Load the program into PIO memory
+    uint offset = pio_add_program(pio, &simple_irq_program);
+
+    // Tell each state machine to run the program.
+    // They are allowed to run the same program in memory.
+    simple_irq_program_init(pio, 0, offset);
+    simple_irq_program_init(pio, 1, offset);
+    simple_irq_program_init(pio, 2, offset);
+    simple_irq_program_init(pio, 3, offset);
+}
+
+/**
+ * Writes to the tx fifo of the given state machine.
+ * This will make the simple_irq program send an ISR to us!
+ */
+void trigger_irq(int sm) {
+    printf("Triggering ISR from state machine %d\n", sm);
+    pio_sm_put_blocking(pio0, sm, 1);
+    // Interrupt will fire from the pio right here since they trigger
+    // whenever data goes into the TX FIFO, see simple_irq.pio
+
+    // Print the irq we expect based on the given state machine
+    printf("Expected IRQ flags: 0x%08X\n", (1 << sm));
+    printf("Actual IRQ Flags: 0x%08X\n", irq_flags);
+
+    // Here you could dispatch work for the irq depending one
+    // flagged it. The work should be done here rather than inside
+    // the irq function. While executing code in an irq, no other
+    // irq's can execute. By having the processing code outside of
+    // the irq, we can still be responsive to new irqs.
+    if (irq_flags & PIO_SM_0_IRQ) {
+        // handle_sm0_irq();
+    }
+    if (irq_flags & PIO_SM_1_IRQ) {
+        // handle_sm1_irq();
+    }
+    if (irq_flags & PIO_SM_2_IRQ) {
+        // handle_sm2_irq();
+    }
+    if (irq_flags & PIO_SM_3_IRQ) {
+        // handle_sm3_irq();
+    }
+    
+    // clear irq flags now.
+    irq_flags = 0;
```

Comment:
Rather than setting _all_ `irq_flags` to 0 here at the end, I wonder if it might be more idiomatic to clear the flags individually e.g. by doing something like `irq_flags &= !PIO_SM_1_IRQ` inside the `if (irq_flags & PIO_SM_1_IRQ) {` block? Or perhaps that's unnecessary complication? :shrug: